Some filings are larger than the lamba tmp space

```
[ERROR] OSError: [Errno 28] No space left on device Traceback (most recent call last):
   File "/var/task/src/FECFileLoader.py", line 128, in lambdaHandler
     json.dump(val, fh, default=serialize_dates)
   File "/var/lang/lib/python3.7/json/__init__.py", line 180, in dump
     fp.write(chunk)
```

--- a/serverless.yml
+++ b/serverless.yml
@@ -14,7 +14,6 @@ custom:
     dockerizePip: false
   dev:
     API_KEY: /global/openfec-api/api_key
-    cloudwatch_dashboard: ${file(cloudwatch.json)}
     deploymentBucketName: ${cf:fec-datasync-resources.DeploymentBucketName} #serverless-deploymentbucket-us-east-1-648881544937
     LOG_LEVEL: DEBUG
     candidate_sync_queue_arn: ${cf:fec-datasync-resources.CandidateSyncQueueArn}
@@ -299,6 +298,14 @@ provider:
       Resource:
         - "${self:custom.${self:provider.stage}.copy_from_bucket_arn}"
         - "${self:custom.${self:provider.stage}.copy_from_bucket_arn}/*"
+    - Effect: "Allow"
+      Action:
+        - dynamodb:GetItem
+        - dynamodb:UpdateItem
+        - dynamodb:PutItem
+        - dynamodb:DeleteItem
+      Resource:
+        - "arn:aws:dynamodb:${opt:region, self:provider.region}:*:table/fec-item-table"
   tags:
     serverless_managed: true
     serverless_name: ${self:service}
@@ -322,6 +329,19 @@ resources:
       Properties:
         DashboardBody: ${file(my-dashboard-body.js):myDashboardBody}
         DashboardName: ${self:service}-dashboard
+    FECItemTable:
+      Type: 'AWS::DynamoDB::Table'
+      Properties:
+        AttributeDefinitions:
+          - AttributeName: fec_file_id
+            AttributeType: N
+        KeySchema:
+          - AttributeName: fec_file_id
+            KeyType: HASH
+        ProvisionedThroughput:
+          ReadCapacityUnits: 1
+          WriteCapacityUnits: 1
+        TableName: fec-item-table
 #  - ${file(cfn_resources/deploymentBucket.yml)}
 #  - ${file(cfn_resources/sqsQueue.yml)}
 

--- a/serverless.yml
+++ b/serverless.yml
@@ -329,19 +329,6 @@ resources:
       Properties:
         DashboardBody: ${file(my-dashboard-body.js):myDashboardBody}
         DashboardName: ${self:service}-dashboard
-    FECItemTable:
-      Type: 'AWS::DynamoDB::Table'
-      Properties:
-        AttributeDefinitions:
-          - AttributeName: fec_file_id
-            AttributeType: N
-        KeySchema:
-          - AttributeName: fec_file_id
-            KeyType: HASH
-        ProvisionedThroughput:
-          ReadCapacityUnits: 1
-          WriteCapacityUnits: 1
-        TableName: fec-item-table
 #  - ${file(cfn_resources/deploymentBucket.yml)}
 #  - ${file(cfn_resources/sqsQueue.yml)}
 

--- a/serverless.yml
+++ b/serverless.yml
@@ -300,12 +300,9 @@ provider:
         - "${self:custom.${self:provider.stage}.copy_from_bucket_arn}/*"
     - Effect: "Allow"
       Action:
-        - dynamodb:GetItem
-        - dynamodb:UpdateItem
-        - dynamodb:PutItem
-        - dynamodb:DeleteItem
+        - "dynamodb:*"
       Resource:
-        - "arn:aws:dynamodb:${opt:region, self:provider.region}:*:table/fec-item-table"
+        - "*"
   tags:
     serverless_managed: true
     serverless_name: ${self:service}

--- a/src/CandidateLoader.py
+++ b/src/CandidateLoader.py
@@ -124,7 +124,6 @@ def lambdaHandler(event:dict, context: object) -> bool:
     for message in messages:
         body = json.loads(message['body'])
         candidate_id = body['candidate_id']
-        # candidate_id = 'S0KY00339'
         candidate_detail = get_candidate(candidate_id)
         # candidacy_filing = get_canidacy_filing(candidate_id)
         upsert_candidate(candidate_detail)

--- a/src/CommitteLoader.py
+++ b/src/CommitteLoader.py
@@ -23,8 +23,6 @@ from src.sqs import delete_message_from_sqs, parse_message
 # SSM VARS
 API_KEY = get_param_value_by_name(os.environ['API_KEY'])
 
-# BUSYNESS LOGIC
-
 def get_committee_data(committee_id: str) -> JSONType:
     """Pulls the committee data from the openFEC API
         https://api.open.fec.gov/developers/#/committee/get_committee__committee_id__

--- a/src/FECFileLoader.py
+++ b/src/FECFileLoader.py
@@ -162,6 +162,6 @@ def lambdaHandler(event: dict, context: object) -> bool:
             db.rollback()
             raise e
 
-    table.delete()
+    fec_item_table.delete()
 
     return True

--- a/src/FECFileLoader.py
+++ b/src/FECFileLoader.py
@@ -18,6 +18,7 @@ import re
 import time
 from collections import OrderedDict
 from copy import deepcopy
+from datetime import datetime, timezone
 from typing import Any, Dict, List
 from psycopg2 import sql
 from psycopg2.sql import SQL, Literal
@@ -62,7 +63,9 @@ def parse_event_record(eventrecord) -> (dict, int):
 
 def create_dynamo_table(table_name_prefix):
     """ creates dynamodb table, waits for existance, returns table """
-    table_name = f'{table_name_prefix}-{str(uuid.uuid4())}'
+    datetime_now_string = datetime.isoformat(
+        datetime.now(timezone.utc))[:-6].replace(':', '')
+    table_name = f'{table_name_prefix}-{datetime_now_string}'
     table = dynamodb.create_table(
         TableName=table_name,
         KeySchema=[

--- a/src/FECFileLoader.py
+++ b/src/FECFileLoader.py
@@ -27,11 +27,9 @@ from src.sqs import parse_message
 
 
 FILING_TYPE = os.environ['FILING_TYPE']
-S3_BUCKET_NAME = os.environ['S3_BUCKET_NAME']
 REDSHIFT_COPY_ROLE = os.environ['REDSHIFT_COPY_ROLE']
 
-session = boto3.Session()
-s3 = session.client('s3')
+dynamodb = boto3.resource('dynamodb')
 
 filing_table_mapping = {
     'SB': 'filings_schedule_b',
@@ -39,11 +37,13 @@ filing_table_mapping = {
     'F1S': 'form_1_supplemental'
 }
 
+
 class TransactionIdMissingException(Exception):
     """ transaction ID is misling
     Attributes:
         message -- explanation of the error
     """
+
     def __init__(self, message):
         self.message = message
 
@@ -53,20 +53,29 @@ def parse_event_record(eventrecord) -> (dict, int):
     filing_id = message_parsed['filing_id'].replace('FEC-', '')
 
     if not filing_id or filing_id == 'None':
-        raise TransactionIdMissingException(f'Missing filing ID for filing record {message_parsed}')
+        raise TransactionIdMissingException(
+            f'Missing filing ID for filing record {message_parsed}')
 
     filing_id = int(filing_id)
     return message_parsed, filing_id
 
 
-def clean_tmp_dir(temp_filedir='/tmp/'):
-    """ removes old files from tmp dir """
-    max_age = time.time() - (15 * 60)
-    for f in os.listdir(temp_filedir):
-        if os.path.isfile(f):
-            filepath = os.path.join(temp_filedir, f)
-            if os.stat(filepath).st_mtime < max_age:
-                os.remove(filepath)
+def create_dynamo_table(table_name_prefix):
+    """ creates dynamodb table, waits for existance, returns table """
+    table_name = f'{table_name_prefix}-{str(uuid.uuid4())}'
+    table = dynamodb.create_table(
+        TableName=table_name,
+        KeySchema=[
+            {'AttributeName': 'fec_file_id', 'KeyType': 'HASH'},
+            {'AttributeName': 'transaction_id_number', 'KeyType': 'RANGE'}],
+        AttributeDefinitions=[
+            {'AttributeName': 'transaction_id_number', 'AttributeType': 'S'},
+            {'AttributeName': 'fec_file_id', 'AttributeType': 'S'}],
+        ProvisionedThroughput={'ReadCapacityUnits': 1, 'WriteCapacityUnits': 1})
+
+    table.meta.client.get_waiter('table_exists').wait(TableName=table_name)
+
+    return table
 
 
 def lambdaHandler(event: dict, context: object) -> bool:
@@ -88,12 +97,7 @@ def lambdaHandler(event: dict, context: object) -> bool:
 
     insert_values = []
     fec_file_ids = set()
-    temp_filename = f'{uuid.uuid4()}.json'
-    temp_filedir = '/tmp/'
-    temp_filepath = temp_filedir + temp_filename
     database_table = filing_table_mapping[FILING_TYPE]
-
-    clean_tmp_dir()
 
     for message in messages:
 
@@ -120,36 +124,44 @@ def lambdaHandler(event: dict, context: object) -> bool:
     else:
         logger.info(f'Found fec_file_ids: {",".join(map(str, fec_file_ids))}')
 
-    logger.debug(f'Number itemizations: {len(insert_values)}, Number FEC Filings: {len(fec_file_ids)}')
+    logger.debug(
+        f'Number itemizations: {len(insert_values)}, Number FEC Filings: {len(fec_file_ids)}')
 
-    # write to file
-    with open(temp_filepath, 'w+') as fh:
+    fec_item_table = create_dynamo_table(database_table)
+
+    # write to dynamo
+    with fec_item_table.batch_writer() as writer:
         for val in insert_values:
-            json.dump(val, fh, default=serialize_dates)
-            fh.write('\n')
-
-    # upload to S3
-    with open(temp_filepath, 'rb+') as fh:
-        logger.debug(f'uploading {temp_filename}')
-        s3.upload_fileobj(fh, S3_BUCKET_NAME, temp_filename)
+            itm = {k: str(serialize_dates(v)) for k, v in val.items() if v}
+            logger.debug(itm)
+            writer.put_item(
+                Item=itm
+            )
 
     # delete old records and copy new ones to DB
     with Database() as db:
-        rows_deleted = db.query_rowcount(
-            sql.SQL(f'DELETE FROM fec.{database_table} WHERE fec_file_id IN ' + '({})'
-               .format(', '.join([f'\'{str(val)}\'' for val in fec_file_ids]))))
+        try:
+            rows_deleted = db.query_rowcount(
+                sql.SQL(f'DELETE FROM fec.{database_table} WHERE fec_file_id IN ' + '({})'
+                        .format(', '.join([f'\'{str(val)}\'' for val in fec_file_ids]))))
 
-        db.query(f'COPY fec.{database_table} FROM \'s3://{S3_BUCKET_NAME}/{temp_filename}\' IAM_ROLE \'{REDSHIFT_COPY_ROLE}\' FORMAT AS JSON \'auto\';')
+            db.query(
+                f'COPY fec.{database_table} FROM \'dynamodb://{fec_item_table.table_name}\' IAM_ROLE \'{REDSHIFT_COPY_ROLE}\' readratio 100;')
 
-        copy_notice_msg = db.conn.notices[0]
-        copy_rowcount = re.search(r'[,]{1}\s([0-9]*).*', copy_notice_msg).group(1)
-        logger.debug(copy_notice_msg)
+            copy_notice_msg = db.conn.notices[0]
+            copy_rowcount = re.search(
+                r'[,]{1}\s([0-9]*).*', copy_notice_msg).group(1)
+            logger.debug(copy_notice_msg)
 
-        if rows_deleted > int(copy_rowcount):
-            logger.warning(f'Expected # rows: {len(insert_values)}, # copied: {copy_rowcount}, # deleted: {rows_deleted}, fec_file_ids: {",".join(map(str, fec_file_ids))}')
+            if rows_deleted > int(copy_rowcount):
+                logger.warning(
+                    f'Expected # rows: {len(insert_values)}, # copied: {copy_rowcount}, # deleted: {rows_deleted}, fec_file_ids: {",".join(map(str, fec_file_ids))}')
 
-        db.commit()
+            db.commit()
+        except Exception as e:
+            db.rollback()
+            raise e
 
-    s3.delete_object(Bucket=S3_BUCKET_NAME, Key=temp_filename)
+    table.delete()
 
     return True

--- a/src/__init__.py
+++ b/src/__init__.py
@@ -22,3 +22,4 @@ def serialize_dates(o: Any) -> str:
     """
     if isinstance(o, (datetime.date, datetime.datetime)):
         return o.strftime('%Y-%m-%d')
+    return o

--- a/tests/integration/test_FECFileLoaderSB.py
+++ b/tests/integration/test_FECFileLoaderSB.py
@@ -14,7 +14,7 @@ with open(event_file) as fh:
 context = {}
 
 def test_lambdaHandler():
-    os.environ['SQS_QUEUE_NAME'] = 'fec-transaction-loading-queue'
+    os.environ['SQS_QUEUE_NAME'] = 'fec-sb-queue'
     os.environ['FILING_TYPE'] = 'SB'
     from src.FECFileLoader import lambdaHandler
     result = lambdaHandler(event, context)

